fix: Add select accessibility refactor v11

davidmenendez opened the pull request:
for #2819 & #2821 

this initial PR addresses keyboard concerns with regards to add select. the main feature being added is arrow key navigation within add select. you should now be able to scroll items with up and down as well as traverse the hierarchy with the right arrow key.

this also breaks up `AddSelectList` into several new components including `AddSelectRow` and `AddSelectFormControl`. the main concern with add select is the usage of native checkboxes and radio buttons. with `AddSelectFormControl` the form controls are now svg's.

elycheea commented on packages/cloud-cognitive/src/components/AddSelect/AddSelectFormControl.js:
```diff
@@ -0,0 +1,90 @@
+} from '@carbon/react/icons';
+import PropTypes from 'prop-types';
+import { pkg } from '../../settings';
+import { UserProfileImage } from '../UserProfileImage';
```
Wonder if we should consider having consumers pass in their own `UserProfileImage` later since this will eventually get deprecated and replaced with `UserAvatar` in v2 ... we might need to support both types? Or we could just update to `UserAvatar` when that time comes. 🤔

elycheea commented on packages/cloud-cognitive/src/components/AddSelect/AddSelectFormControl.js:
```diff
@@ -0,0 +1,90 @@
+    onClick,
+    id: item.id,
+    className: `${blockClass}-form-control-wrapper`,
+    [`aria-label`]: item.title,
```
Should probably add `size: 20` here — icons appearing a bit smaller than in previous versions.

Current | Previous
--------|---------
![image](https://user-images.githubusercontent.com/1717398/230210795-27b34b5a-d252-4513-9371-0abc0c3467bf.png) | ![image](https://user-images.githubusercontent.com/1717398/230211293-5e734b04-20ea-4b69-9a23-e506a1562317.png)

Or was this a conscious decision to use smaller checkboxes since the icons have a thicker stroke at this size also still a bit smaller. 👀 
![image](https://user-images.githubusercontent.com/1717398/230211424-9438e431-954d-4858-a665-211f2785449e.png)

elycheea commented on packages/cloud-cognitive/src/components/AddSelect/AddSelectFormControl.js:
```diff
@@ -0,0 +1,90 @@
+
+  const getCheckbox = () => {
+    if (selected) {
+      return <CheckboxChecked {...controlProps} />;
```
```suggestion
      return <CheckboxCheckedFilled {...controlProps} />;
```

davidmenendez commented:
there's still a few kinks to work out before this is ready. i'll be working on this today.

davidmenendez commented:
there's still a few kinks to work out before this is ready. i'll be working on this today.

davidmenendez commented:
@elycheea i addressed most of your concerns and stuff you pointed out. the only exception being multi add select with hierarchy. that's going to require some special consideration and i've already got an issue open specifically to address that.

elycheea commented on packages/cloud-cognitive/src/components/AddSelect/AddSelectRow.js:
```diff
@@ -0,0 +1,227 @@
+  };
+
+  const onSelectKeyDown = ({ key }) => {
+    if (key === 'Enter') {
```
Hopefully a quick fix, but can we get this working with both <kbd>Enter</kbd> and <kbd>Space</kbd>? I noticed that the actual checkboxes in MultiAddSelect require `space` so this inconsistency feels unexpected.

elycheea reviewed (approved):
Approved for improvements to accessibility.

Things to be addressed later, continued improvements to `MultiAddSelect`, `UserProfileImage`/`UserAvatar` (future) handling, surprise updates to profile image background colors. As discussed, these will be logged and addressed separately.
